hw/mcu/nxp/kinetis/src/hal_i2c.c

```diff
@@ -0,0 +1,408 @@
+#include "mcu/kinetis_hal.h"
+#include "mcu/mcu.h"
+#include "hal/hal_i2c.h"
+
+#include <fsl_i2c.h>
+#include <fsl_port.h>
+#include <fsl_clock.h>
+
+#define NXP_HAL_I2C_MAX 4
+
+struct nxp_hal_i2c {
+    I2C_Type *dev;
+    uint32_t scl_pin;
+    uint32_t sda_pin;
+    PORT_Type *port;
+    port_mux_t mux;
+    IRQn_Type irqn;
+    void (*irq_handler)(void);
+    i2c_master_handle_t handle;
+    struct os_sem sync;
+    status_t stat;
+    bool enabled;
+    bool ongoing;
+};
+
+#if MYNEWT_VAL(I2C_0)
+static void i2c0_irq(void);
+static struct nxp_hal_i2c hal_i2c0 = {
+    .dev = I2C0,
+    .scl_pin = MYNEWT_VAL(I2C_0_PIN_SCL),
+    .sda_pin = MYNEWT_VAL(I2C_0_PIN_SDA),
+    .port = MYNEWT_VAL(I2C_0_PORT),
+    .mux = MYNEWT_VAL(I2C_0_MUX),
+    .irqn = I2C0_IRQn,
+    .irq_handler = i2c0_irq,
+    .stat = 0,
+    .enabled = false
+};
+#endif
+
+#if MYNEWT_VAL(I2C_1)
+static void i2c1_irq(void);
+static struct nxp_hal_i2c hal_i2c1 = {
+    .dev = I2C1,
+    .scl_pin = MYNEWT_VAL(I2C_1_PIN_SCL),
+    .sda_pin = MYNEWT_VAL(I2C_1_PIN_SDA),
+    .port = MYNEWT_VAL(I2C_1_PORT),
+    .mux = MYNEWT_VAL(I2C_1_MUX),
+    .irqn = I2C1_IRQn,
+    .irq_handler = i2c1_irq,
+    .stat = 0,
+    .enabled = false
+};
+#endif
+
+#if MYNEWT_VAL(I2C_2)
+static void i2c2_irq(void);
+static struct nxp_hal_i2c hal_i2c2 = {
+    .dev = I2C2,
+    .scl_pin = MYNEWT_VAL(I2C_2_PIN_SCL),
+    .sda_pin = MYNEWT_VAL(I2C_2_PIN_SDA),
+    .port = MYNEWT_VAL(I2C_2_PORT),
+    .mux = MYNEWT_VAL(I2C_2_MUX),
+    .irqn = I2C2_IRQn,
+    .irq_handler = i2c2_irq,
+    .stat = 0,
+    .enabled = false
+};
+#endif
+
+#if MYNEWT_VAL(I2C_3)
+static void i2c3_irq(void);
+static struct nxp_hal_i2c hal_i2c3 = {
+    .dev = I2C3,
+    .scl_pin = MYNEWT_VAL(I2C_3_PIN_SCL),
+    .sda_pin = MYNEWT_VAL(I2C_3_PIN_SDA),
+    .port = MYNEWT_VAL(I2C_3_PORT),
+    .mux = MYNEWT_VAL(I2C_3_MUX),
+    .irqn = I2C3_IRQn,
+    .irq_handler = i2c3_irq,
+    .stat = 0,
+    .enabled = false
+};
+#endif
+
+static struct nxp_hal_i2c *i2c_modules[NXP_HAL_I2C_MAX] = {
+#if MYNEWT_VAL(I2C_0)
+    &hal_i2c0,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(I2C_1)
+    &hal_i2c1,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(I2C_2)
+    &hal_i2c2,
+#else
+    NULL,
+#endif
+#if MYNEWT_VAL(I2C_3)
+    &hal_i2c3,
+#else
+    NULL,
+#endif
+};
+
+#if MYNEWT_VAL(I2C_0)
+static void
+i2c0_irq(void)
+{
+    I2C_MasterTransferHandleIRQ(hal_i2c0.dev, &hal_i2c0.handle);
+}
+#endif
+#if MYNEWT_VAL(I2C_1)
+static void
+i2c1_irq(void)
+{
+    I2C_MasterTransferHandleIRQ(hal_i2c1.dev, &hal_i2c1.handle);
+}
+#endif
+#if MYNEWT_VAL(I2C_2)
+static void
+i2c2_irq(void)
+{
+    I2C_MasterTransferHandleIRQ(hal_i2c2.dev, &hal_i2c2.handle);
+}
+#endif
+#if MYNEWT_VAL(I2C_3)
+static void
+i2c3_irq(void)
+{
+    I2C_MasterTransferHandleIRQ(hal_i2c3.dev, &hal_i2c3.handle);
+}
+#endif
+
+static void
+master_xfer_cb(I2C_Type *dev,
+               i2c_master_handle_t *handle,
+               status_t status,
+               void *userData)
```

Comment:
Naming convention: should be `user_data`

Reply:
Addressed.